annotation/src/main/java/org/jenkinsci/Symbol.java

```diff
@@ -45,4 +45,5 @@
 @Documented
 public @interface Symbol {
     String[] value();
+    Class<?>[] context() default {};
```

Comment:
I think I like this idea.

Would help to express your plan in the javadoc for this.

I think what you want is that where there are multiple options with the same name, if there is one with the "closest" context, then that one will win.

TBH I like that, because we can do something like `@Symbol(value="discoverOriginPRs", context={GitHubSCMSource.class,GitHubSCMNavigator.class})` and we get the win.

The problem, as I see it, is in deciding which is the *nearer* context if the context is not an exact match.

i.e. If the context is an interface or a parent class.

You'll need to clearly spell out how you decide the nearest context... and if multiple symbols are the same name at the same distance, we fall back to ambiguous

Reply:
My original idea was to just have an extension point that the symbol library could consult in the case where it had conflicts.

The extension point would basically be given a list of candidate symbols and the context in which these were requested and remove any that it wants from the list.

We then walk all implementations of the extension in order. If the list is empty, we stop and say "no match". If we get to the end and there is more than one, we stop and say "ambiguous" and if we have exactly one "winner"

And we need to process the list of extensions all the way to the end, as we don't know if a later impl might remove the one option we are left with bringing us to an empty list

Reply:
But with this... while requiring changes on the implementation plugins... we get a less "magic" solution... so as long as you define what is "closest" I think this will work really well